File: windows/src/RealtimeTranslator.Core/Subtitles/SubtitleTailClipper.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace RealtimeTranslator.Core.Subtitles;

/// <summary>字幕表示用に末尾 N 文字だけを残す。</summary>
public static class SubtitleTailClipper
{
    public const int JapaneseCharacterLimit = 60;
    public const int EnglishCharacterLimit = 120;
    public const string Ellipsis = "…";

    public static string Clip(string text)
    {
        ArgumentNullException.ThrowIfNull(text);

        var trimmed = text.Trim();
        if (trimmed.Length == 0)
        {
            // 空白のみの入力は呼び出し元の空判定を壊さないようそのまま返す。
            return text;
        }

        var isCjk = ContainsCjk(trimmed);
        var limit = isCjk ? JapaneseCharacterLimit : EnglishCharacterLimit;
        var elements = TextElements(trimmed);
        if (elements.Count <= limit)
        {
            return trimmed;
        }

        var suffix = string.Concat(elements.GetRange(elements.Count - limit, limit));
        var tail = isCjk ? suffix : DropLeadingPartialWord(suffix);
        return tail.Length == 0 ? Ellipsis + suffix : Ellipsis + tail;
    }

    public static bool ContainsCjk(string text)
    {
        ArgumentNullException.ThrowIfNull(text);

        foreach (var rune in text.EnumerateRunes())
        {
            if (rune.Value is (>= 0x3040 and <= 0x30FF) or (>= 0x4E00 and <= 0x9FFF) or (>= 0x3400 and <= 0x4DBF))
            {
                return true;
            }
        }

        return false;
    }

    /// <summary>英語の単語途中切れを避けるため、先頭から最初の空白までを捨てる。</summary>
    private static string DropLeadingPartialWord(string text)
    {
        for (var index = 0; index < text.Length; index++)
        {
            if (!char.IsWhiteSpace(text[index]))
            {
                continue;
            }

            var afterSpace = index + 1;
            return afterSpace < text.Length ? text[afterSpace..] : text;
        }

        return text;
    }

    /// <summary>Swift の Character 単位カウントに合わせて書記素クラスタで分割する。</summary>
    private static List<string> TextElements(string text)
    {
        var elements = new List<string>(text.Length);
        var offset = 0;
        while (offset < text.Length)
        {
            var length = StringInfo.GetNextTextElementLength(text.AsSpan(offset));
            elements.Add(text.Substring(offset, length));
            offset += length;
        }

        return elements;
    }
}

File: RealtimeTranslator/Subtitles/SubtitleTailClipper.swift
import Foundation

/// 字幕表示用に末尾N文字だけを残す。SwiftUIの head truncation は
/// 1行目を先頭固定にしたまま2行目だけを省略するため、表示前に切り詰める。
enum SubtitleTailClipper {
    static let japaneseCharacterLimit = 60
    static let englishCharacterLimit = 120
    static let ellipsis = "…"

    static func clip(_ text: String) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return text }

        let limit =
            containsCJK(trimmed)
            ? japaneseCharacterLimit
            : englishCharacterLimit
        guard trimmed.count > limit else { return trimmed }

        var tail = String(trimmed.suffix(limit))
        if !containsCJK(trimmed) {
            tail = dropLeadingPartialWord(from: tail)
        }
        guard !tail.isEmpty else {
            return ellipsis + String(trimmed.suffix(limit))
        }
        return ellipsis + tail
    }

    static func containsCJK(_ text: String) -> Bool {
        text.unicodeScalars.contains { scalar in
            (0x3040...0x30FF).contains(scalar.value)
                || (0x4E00...0x9FFF).contains(scalar.value)
                || (0x3400...0x4DBF).contains(scalar.value)
        }
    }

    /// 英語の単語途中切れを避けるため、先頭から最初の空白までを捨てる。
    private static func dropLeadingPartialWord(from text: String) -> String {
        guard let firstSpace = text.firstIndex(where: { $0.isWhitespace }) else {
            return text
        }
        let afterSpace = text.index(after: firstSpace)
        guard afterSpace < text.endIndex else { return text }
        return String(text[afterSpace...])
    }
}
